pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,195 @@
+
+        def jacobian_wrapper(*args, **kwargs):
+            qjac = _wrapper(*args, **kwargs)
+
+            if any(m.return_type is qml.operation.Probability for m in qnode.qtape.measurements):
+                qjac = qml.math.squeeze(qjac)
+
+            if not hybrid:
+                return qjac
+
+            kwargs.pop("shots", False)
+            cjac = cjac_fn(*args, **kwargs)
+
+            if isinstance(cjac, tuple):
+                # Classical processing of multiple arguments is present. Return qjac @ cjac.
+                jacs = [
+                    qml.math.squeeze(qml.math.tensordot(c, qjac, [[0], [-1]]))
+                    for c in cjac
+                    if c is not None
+                ]
+                return jacs
+
+            is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
+
+            if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
+                # Classical Jacobian is the identity. No classical processing
+                # is present inside the QNode.
+                return qjac
+
+            # Classical processing of a single argument is present. Return cjac.T @ qjac @ cjac.
+            jac = qml.math.tensordot(qjac, cjac, [[-1], [0]])
+            jac = qml.math.tensordot(jac, cjac, [[-2], [0]])
+            return qml.math.swapaxes(jac, -1, -2)
+
+        return jacobian_wrapper
+
+
+@hessian_transform
+def param_shift_hessian(tape):
+    """Generate parameter-shift tapes and postprocessing method
+    to directly compute the hessian."""
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    _gradient_analysis(tape)
+    gradient_tapes = []
+    gradient_coeffs = []
+    shapes = []
+    dim = len(tape.trainable_params)
+
+    if not tape.trainable_params:
+        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    # The Hessian for a 2-term parameter-shift rule can be expressed via the following recipes.
+    # Off-diagonal elements of the Hessian require shifts to two different parameter indices.
+    # A recipe can thus be expressed via the tape patterns: (dummy values for ndarray creation)
+    #       [[coeff, dummy], [mult, dummy], [shift1, shift2]]
+    # Each corresponding to one term in the parameter-shift formula:
+    #       didj f(x) = coeff * f(mult*x + shift1*ei +shift2*ej) + ...
+    diag_recipe = [
+        [[ 0.5], [1], [np.pi]],
+        [[-0.5], [1], [    0]]
+    ]
+    off_diag_recipe = [
+        [[ 0.25, 1], [1, 1], [ np.pi/2,  np.pi/2]],
+        [[-0.25, 1], [1, 1], [-np.pi/2,  np.pi/2]],
+        [[-0.25, 1], [1, 1], [ np.pi/2, -np.pi/2]],
+        [[ 0.25, 1], [1, 1], [-np.pi/2, -np.pi/2]]
+    ]
```

Comment:
Makes sense to hardcode it for now 👍 

I wonder if we could programmatically determine the 2nd order shift rule given an arbitrary 1st order shift rule? Perhaps if we could encode the 1st order rule as a function representing a linear combination, we would be able to do `qml.jacobian(grad_recipe)(shift)`?